polkadot/primitives/src/vstaging/mod.rs

```diff
@@ -98,3 +116,873 @@ impl<BlockNumber: Default + From<u32>> Default for SchedulerParams<BlockNumber>
+		if self.version() == CandidateDescriptorVersion::V1 {
+			return Some(self.rebuild_signature_field())
+		}
+
+		None
+	}
+
+	/// Returns the `core_index` of `V2` candidate descriptors, `None` otherwise.
+	pub fn core_index(&self) -> Option<CoreIndex> {
+		if self.version() == CandidateDescriptorVersion::V1 {
+			return None
+		}
+
+		Some(CoreIndex(self.core_index as u32))
+	}
+
+	/// Returns the `core_index` of `V2` candidate descriptors, `None` otherwise.
+	pub fn session_index(&self) -> Option<SessionIndex> {
+		if self.version() == CandidateDescriptorVersion::V1 {
+			return None
+		}
+
+		Some(self.session_index)
+	}
+}
+
+impl CommittedCandidateReceiptV2 {
+	/// Performs a sanity check of the descriptor and commitment.
+	///
+	/// Returns error if:
+	/// - descriptor core index is different than the core selected
+	/// by the commitments
+	/// - the core index is out of bounds
+	pub fn check(
+		&self,
+		n_cores: u32,
+		// TODO: consider promoting `ClaimQueueSnapshot` as primitive
+		claim_queue: &BTreeMap<CoreIndex, VecDeque<ParaId>>,
+	) -> Result<(), CandidateReceiptError> {
+		// Don't check v1 descriptors.
+		if self.descriptor.version() == CandidateDescriptorVersion::V1 {
+			return Ok(())
+		}
+
+		if claim_queue.is_empty() {
+			return Err(CandidateReceiptError::NoAssignment)
+		}
+
+		let descriptor_core_index = CoreIndex(self.descriptor.core_index as u32);
+		if descriptor_core_index.0 > n_cores - 1 {
+			return Err(CandidateReceiptError::InvalidCoreIndex)
+		}
+
+		let (core_selector, cq_offset) =
+			self.commitments.selected_core().ok_or(CandidateReceiptError::NoCoreSelected)?;
+
+		// TODO: double check the invariant of claim queue len == scheduling lookahead.
+		// What happens when on-demand is used and there is only 1 claim on a core.
+		let claim_queue_depth = claim_queue
+			.first_key_value()
+			.ok_or(CandidateReceiptError::NoAssignment)?
+			.1
+			.len();
+		if cq_offset.0 as usize >= claim_queue_depth {
+			return Err(CandidateReceiptError::InvalidSelectedCore)
+		}
+
+		let para_id = self.descriptor.para_id;
+
+		// Get a vec of the core indices the parachain is assigned to at `cq_offset`.
+		let assigned_cores = claim_queue
+			.iter()
+			.filter_map(|(core_index, queue)| {
+				let queued_para = queue.get(cq_offset.0 as usize)?;
+
+				if queued_para == &para_id {
+					Some(core_index)
+				} else {
+					None
+				}
+			})
+			.collect::<Vec<_>>();
+
+		if assigned_cores.is_empty() {
+			return Err(CandidateReceiptError::NoAssignment)
+		}
+
+		let core_index = assigned_cores
+			.get(core_selector.0 as usize % assigned_cores.len())
+			.expect("provided index is always less than queue len; qed");
+
+		if **core_index != descriptor_core_index {
+			return Err(CandidateReceiptError::CoreIndexMismatch)
+		}
+
+		Ok(())
+	}
+}
+
+/// A backed (or backable, depending on context) candidate.
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+pub struct BackedCandidate<H = Hash> {
```

Comment:
Why do we need to redefine all of these, if the new format is binary compatible? We should be able to directly replace the old receipt type with the new one, right?

Reply:
Yeah, probablythis is not the right place to put these in. There are two reasons to keep these duplicated:
- We still want collators and validators to check the collator signature so we'd have to keep the old candidate receipt.
- I don't want to change the node side in this PR. 

When we change the node side to the new primitives, it will still be using the old receipts until we enable the feature bit.

Reply:
> We still want collators and validators to check the collator signature so we'd have to keep the old candidate receipt.

hmm but since they're binary compatible, one may just convert the newly-added fields back to their original meaning (you already added methods for that) when needed.

If we did that (and changed the node side in this PR), we could remove a lot of duplicated definitions.

Reply:
Yeah, certainly this will be done in the PR that adds support for these primitives on the node. I don’t really see a problem keeping them until then.